src/phl/ecPhlCodeTx.ml

```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
   t_code_transform side ~bdhoare:true cpos tr (t_zip (set_stmt (fresh, id) e)) tc
 
 (* -------------------------------------------------------------------- *)
-let cfold_stmt (pf, hyps) me olen zpr =
+let cfold_stmt ?(simplify=true) (pf, hyps) (me: memenv) olen zpr =
   let env = LDecl.toenv hyps in
 
-  let (asgn, i, tl) =
-    match zpr.Zpr.z_tail with
-    | ({ i_node = Sasgn (lv, e) } as i) :: tl -> begin
-      let asgn =
-        match lv with
-        | LvVar (x, ty) -> [(x, ty, e)]
-        | LvTuple xs -> begin
-            match e.e_node with
-            | Etuple es -> List.map2 (fun (x, ty) e -> (x, ty, e)) xs es
-            | _ -> assert false
-        end
-      in
-        (asgn, i, tl)
-    end
+  (* Fold constant until next assignment or end of given # of lines *)
+  let rec doit 
+    ((asg, subst): PV.t * (expr,  unit) Mpv.t) 
+    (olen: int option) 
+    (acc: instr list) 
+    (insts: instr list) 
+    : instr list =
+    let (asgn, tl) =
+      match insts with
+      | ({ i_node = Sasgn (lv, e) }) :: tl -> begin
+        let asgn =
+          match lv with
+          | LvVar (x, ty) -> 
+            (* check if first iteration or if variable is being propagated *)
+            let () = assert (asg = PV.empty || PV.mem_pv env x asg) in
+            [(x, ty, e)]
+          | LvTuple xs -> begin
+              match e.e_node with
+              | Etuple es -> 
+                (* check if first iteration or if all the variables in the tuple are being propagated *)
+                let () = assert (asg = PV.empty || 
+                    List.for_all (fun x -> PV.mem_pv env x asg) (List.fst xs)) in
+                List.map2 (fun (x, ty) e -> (x, ty, e)) xs es
+              | _ -> assert false
+          end
+        in
+          (asgn, tl)
+      end
+
+      | _ -> 
+          tc_error pf "cannot find a left-value assignment at given position"
+    in
 
-    | _ ->
-        tc_error pf "cannot find a left-value assignment at given position"
-  in
+    (* Check that we are assigning to local variables only *)
+    List.iter
+      (fun (x, _, _) ->
+        if is_glob x then
+          tc_error pf "left-values must be local variables")
+      asgn;
+
+    (* Check that we are assigning an expression that we can evaluate to a constant  *)
+    let expr_is_closed (e: expr) : bool = e_fv e = Mid.empty && e_read env e = PV.empty in
```

Comment:
Use `Mid.is_empty` (I am not sure there is a guarantee that all empty maps are structurally equal to `Mid.empty`)

Same for `PV.empty` (add a `PV.is_empty` function and use it)

Reply:
There is already a PV.is_empty function